### ISSUE TITLE
Multithreading

### ISSUE THREAD

**Issue by EtienneBruines** [paked/engi#53]
As Go is such a nice language to do multithreading in, and most games lack multithreading, it'd be nice if we supported it. 

Currently, each frame, we are iterating over all `System`s, and iterating over all `Entity`s "in" those `System`s. 

I'm not sure if we can safely multithread multiple `System`s at once, but we could easily multithread the iteration over `Entity`s?

### Current implementation
```go
    for _, system := range w.Systems() {
		if headless && system.SkipOnHeadless() {
			continue // so skip it
		}

		system.Pre()
		for _, entity := range system.Entities() {
			if w.paused {
				ok := entity.GetComponent(&unp)
				if !ok {
					continue // so skip it
				}
			}
			if entity.Exists {
				system.Update(entity, dt)
			}
		}
		system.Post()
	}
```

### Proposed implementation
```go
    for _, system := range w.Systems() {
		if headless && system.SkipOnHeadless() {
			continue // so skip it
		}

		entities := system.Entities()

		system.Pre()
                
        // It's not always faster to multithread; so in this case we're not going to
		if len(entities) < 2*runtime.NumCPU() {
			for _, ent := range entities {
				system.Update(ent, dt)
			}
            system.Post()
			continue // with other Systems
		}

		entityChannel := make(chan *Entity)
		wg := sync.WaitGroup{}

		// Launch workers
		for i := 0; i < runtime.NumCPU(); i++ {
			go func() {
				for ent := range entityChannel {
					system.Update(ent, dt)
					wg.Done()
				}
			}()
		}

        // Give them something to do
		for _, entity := range entities {
			if w.paused {
				ok := entity.GetComponent(&unp)
				if !ok {
					continue // so skip it
				}
			}
			if entity.Exists {
				wg.Add(1)
				entityChannel <- entity
			}
		}

        // Wait until they're done, before continuing to other Systems
		wg.Wait()
		close(entityChannel)

		system.Post()
	}
```

I know it's a bit more code than the original implementation, but it sure is efficient.

**Comment by matiwinnetou** [comment]
This is quite controversial change. Keep in mind that there are issues to write from multiple threads to the same components so all components data access would have to be synchronized.

If we would be able to solve this problem we would rock but problem is rather tricky IMHO.

Putting read write mutex everywhere may not buy us much either. For 2D games which this engines aims for this is an extra complexity, which is not perhaps necessary albeit very awesome from technical point of view.

**Comment by EtienneBruines** [comment]
That depends; how often do `System`s write to something else than the object it's currently handling? 
Are there any aspects in particular where you expect to see problems? (multithreading in Go is quite easy, so I think it could be done). 

Keep in mind that we're not letting two `System`s process in parallel. Only one `System` would be handling all `Entity`s at the same time.

**Comment by paked** [comment]
There are actually a *boat* load of problems with using concurrency with Go in my experience. If we cannot guarantee that a group of systems the game engine will actually behave differently.

Consider a collision system and movement system. Depending on the order these are ran, they will have ***different*** game-breaking outcomes.

> `CollisionSystem` to `MovementSystem`

Leaves the player falling into the ground when the state is rendered, this is because the collision is accounted for... And *then* it the player is moved to account for gravity.

> `MovementSystem` to `CollisionSystem`

This is the correct order, gravity and movement is calculated and then if it collides the player is moved again.

Maybe I am misunderstanding the use of your concurrency. In that case, correct me.

**Comment by EtienneBruines** [comment]
I get your example. But:
* this is a problem even without multithreading; 
    * As they are now also handled `System A`, then `System B`
* this could be fixed, even when using multithreading.
    * All we need is some kind of `Priority` for the `System`s.

**Comment by EtienneBruines** [comment]
Is it OK if I spend some time on this, to see if we can do this without changing the *current* behavior? Then when reviewing the PR, we could always say GO/NO-GO. 

(if you think it's a waste of time: I'll probably be starting in 8 hours or so; and doing it for fun - so I won't be offended if it's denied afterwards)

**Comment by paked** [comment]
@EtienneBruines If you can make it work, I am all ears. Go for it!

**Comment by EtienneBruines** [comment]
@matiwinnetou I used your `galaxian` example, just to show it works for massive numbers of renderables. 

Note that it becomes slower over time, because of #56. This is expected. 

Also note that the lagging starts after it starts to render 60.000 objects - this is not shown, though. 

@paked @matiwinnetou You can look at the code; a bit more complicated, but then again: once it works, it works. And I believe it works. Thoughts?

**Comment by paked** [comment]
This is pretty awesome. I'll review this.

**Comment by paked** [comment]
@EtienneBruines now that we have benchmarks, would you mind them on this?

**Comment by EtienneBruines** [comment]
What kind of benchmarks are you looking for? Anything in particular? (the estimate of 60k objects in the post above, was based on that benchmark implementation)

**Comment by paked** [comment]
@EtienneBruines A comparison between non-concurrent vs concurrent would be good

**Comment by EtienneBruines** [comment]
OK. Will do.

**Comment by EtienneBruines** [comment]
I am amazed by the results:

```
BenchmarkWorldWorkersOne-4  	  100000	     55365 ns/op
BenchmarkWorldWorkersTwo-4  	  100000	     49654 ns/op
BenchmarkWorldWorkersFour-4 	  100000	     49153 ns/op
BenchmarkWorldWorkersEight-4	  100000	     60911 ns/op
BenchmarkWorldLegacy-4      	  100000	     38713 ns/op
```

I truly expected a different outcome. I'm investigating as to how this is even possible.

**Comment by EtienneBruines** [comment]
Seeing the results above, I think we could create a per-system setting of `RunInParallel`? Depending on the workload of the `System`, the user could say whether or not he wants to run it in parallel; because it may either greatly improve performance, or greatly reduce.

Thoughts @paked @matiwinnetou ?

**Comment by paked** [comment]
I'd like to see this proposed... Not quite sure how reliable it would be in
practice.

**Comment by EtienneBruines** [comment]
Well, `Systemer` is an interface which defines everything a `System` needs to have. It would include an additional:

```go
type Systemer interface {
    RunInParallel() bool
}
```

Indicating whether or not it should process all `Entity`s in parallel. The `System` struct would return `false` just to be safe, and people can set it to `true` whenever they feel like it would boost their performance. 

This is desired, probably because it's the only way they'd be multithreading at all - no other way to tell `engi`: "give me all `Entity`s at multiple goroutines".

**Comment by paked** [comment]
OK. I understand now. This seems like it would be totally kick ass.

**Comment by matiwinnetou** [comment]
@EtienneBruines  

"So, depending on how fast the Update function on the System is, it makes perfect sense to multitask, or it will only make it slower." 

This does not surprise me at all, matches with knowledge I have based on how multithreaded systems work.

**Comment by EtienneBruines** [comment]
The turning point is somewhere between 50-500 nanoseconds; not sure what's the "average time" we a `System` needs.

**Comment by Kunde21** [comment]
If we can be sure that a call to `System.Update(entity *Entity, dt float32)` won't cause any internal race conditions, there's no reason to limit the number of go-routines in the concurrent implementation.  

    func (w *World) update(dt float32) {
            w.pre()

            var unp *UnpauseComponent

            for _, system := range w.Systems() {
                    if headless && system.SkipOnHeadless() {
                            continue // so skip it
                    }

                    system.Pre()
                    entities := system.Entities()
                    complChan := make(chan struct{})
                    count := len(entities)
                    for _, entity := range entities {
                            if (w.paused && !entity.Component(&unp)) || !entity.Exists {
                                    count--
                                    continue // so skip it
                            }
                            go func() {
                                    system.Update(entity, dt)
                                    complChan <- struct{}{}
                            }()
                    }
                    for i := 0; i < count; i++ {
                            <-complChan
                    }
                    system.Post()
        }
        w.post()
    }

The runtime scheduler is highly optimized and go-routine creation overhead is much less than a hardware thread, so there's no need to limit it to one per CPU core.

**Comment by EtienneBruines** [comment]
Is there a reason not to limit it? (i.e.: with Intels HyperThreading, one could limit to `2*NumCPU`). 

Even if there's little overhead, what's the reason for having the overhead, if it can be avoided? 

(I don't mean to "attack" you; just curious - I might even learn something)

**Comment by Kunde21** [comment]
Limiting the number of workers using channel constructs means you're paying the time and resource overhead to synchronize and pass data via the channel.  There's no guarantee that the dispatcher is able to saturate the workers with tasks, resulting in under-utilized worker go-routines.  

Spinning off a go-routine per job means that every go-routine has 100% utilization, internally, and the runtime scheduler will efficiently multiplex the work across hardware threads.  

There was in interesting [tweet](https://twitter.com/bradfitz/status/661881981995769857) about this a couple weeks ago.

**Comment by EtienneBruines** [comment]
Interesting that you mention that tweet; I also noticed it, but also noticed that neither of those things he's comparing, are actually executed in parallel. If one uses multiple long-lived workers, the workers are actually faster. 

But I only tested that hypothesis using the same example as bradfitz (multiplying the input by `2`). Not sure if it holds in this scenario. 

The up-side about this, is that we don't need an output-channel. All we need, is know that everything has been processed. (that would further simplify your snippet).

**Comment by Kunde21** [comment]
The `complChan` channel in my snippet is used in place of the waitgroup to ensure all go-routines are executed before moving on to `system.Post()`.  Sending a `struct{}{}` down a channel doesn't allocate any memory, so the cost of that operation is minimal.

**Comment by EtienneBruines** [comment]
My bad; you're completely right.

**Comment by Kunde21** [comment]
Do you have those benchmarks handy from (these tests)[https://github.com/paked/engi/issues/53#issuecomment-154842634]?

**Comment by Kunde21** [comment]
I put a few more steps into the current benchmarks.   It seems that the tipping point is around 20 entities in a system.  The efficiency of concurrency grows really quickly above there, to 2-4x speed-up at 1,000 entities.  

Here's my results for `GOMAXPROCS` set at 8 and 4 cores:  
```
Concurrent-8:
BenchmarkCollisionSystem10-8       30000             51,234 ns/op            1,905 B/op         86 allocs/op
BenchmarkCollisionSystem15-8       20000             85,077 ns/op            3,649 B/op        137 allocs/op
BenchmarkCollisionSystem20-8       10000            125,822 ns/op            5,986 B/op        216 allocs/op
BenchmarkCollisionSystem25-8       10000            182,450 ns/op            8,929 B/op        287 allocs/op
BenchmarkCollisionSystem50-8        3000            421,889 ns/op           33,452 B/op        906 allocs/op
BenchmarkCollisionSystem100-8       2000          1,059,544 ns/op          134,764 B/op       3,057 allocs/op
BenchmarkCollisionSystem1000-8        20         85,438,450 ns/op       12,417,864 B/op     265,484 allocs/op

Concurrent-4:
BenchmarkCollisionSystem10-4       30000             49,635 ns/op            1,904 B/op         86 allocs/op
BenchmarkCollisionSystem15-4       20000             86,418 ns/op            3,648 B/op        137 allocs/op
BenchmarkCollisionSystem20-4       10000            127,601 ns/op            5,985 B/op        216 allocs/op
BenchmarkCollisionSystem25-4       10000            172,231 ns/op            8,930 B/op        287 allocs/op
BenchmarkCollisionSystem50-4        3000            431,928 ns/op           33,453 B/op        906 allocs/op
BenchmarkCollisionSystem100-4       1000          1,253,998 ns/op          134,770 B/op       3,058 allocs/op
BenchmarkCollisionSystem1000-4        10        127,351,280 ns/op       12,537,835 B/op     273,501 allocs/op
```

```
Serial-8:
BenchmarkCollisionSystem10-8       50000             39,582 ns/op            1,696 B/op         83 allocs/op
BenchmarkCollisionSystem15-8       20000             79,311 ns/op            3,440 B/op        134 allocs/op
BenchmarkCollisionSystem20-8       10000            134,940 ns/op            5,776 B/op        213 allocs/op
BenchmarkCollisionSystem25-8        5000            210,266 ns/op            8,720 B/op        284 allocs/op
BenchmarkCollisionSystem50-8        2000            804,187 ns/op           33,237 B/op        903 allocs/op
BenchmarkCollisionSystem100-8        500          3,406,920 ns/op          134,604 B/op       3,058 allocs/op
BenchmarkCollisionSystem1000-8         3        360,511,833 ns/op       13,342,314 B/op     324,384 allocs/op

Serial-4
BenchmarkCollisionSystem10-4       50000             38,015 ns/op            1,696 B/op         83 allocs/op
BenchmarkCollisionSystem15-4       20000             75,529 ns/op            3,440 B/op        134 allocs/op
BenchmarkCollisionSystem20-4       10000            129,650 ns/op            5,776 B/op        213 allocs/op
BenchmarkCollisionSystem25-4        5000            204,636 ns/op            8,720 B/op        284 allocs/op
BenchmarkCollisionSystem50-4        2000            834,539 ns/op           33,237 B/op        903 allocs/op
BenchmarkCollisionSystem100-4        500          3,241,025 ns/op          134,592 B/op       3,057 allocs/op
BenchmarkCollisionSystem1000-4         3        353,174,266 ns/op       13,292,928 B/op     321,298 allocs/op```

**Comment by Kunde21** [comment]
Added PR #88

**Comment by Kunde21** [comment]
I just realized the misunderstanding in your comment here:  https://github.com/paked/engi/issues/53#issuecomment-158683322.  In fact, that was the same thing I initially thought.  

That benchmark code isn't comparing concurrent to serial.  Instead, it's comparing the cost of passing a value between go-routines through a channel to passing a value between go-routines via function call to a closure.  

It's a comparison, essentially, of the overhead costs that we were discussing in:  https://github.com/paked/engi/issues/53#issuecomment-158680642 and https://github.com/paked/engi/issues/53#issuecomment-158683005.  That's why I said spawning off a go-routine is cheaper than passing a value via channel to a worker.  

It's probably the biggest break when moving from c-type threads to go-routines (my coding path).  Since the runtime absorbs the cost of creating a hardware thread, we come out ahead when we feed enough go-routines down that hardware thread.  Some webservers have logged millions of concurrent go-routines, which spreads the cost of the hardware thread across all of the go-routines that use it.  

My benchmark results are indicative of using go-routines to spread re-entrant functions across threads.  Native support for map-reduce isn't implemented in go.  That, in my opinion, is mainly due to the go-routine pattern that I'm describing, which accomplishes the map/reduce pattern with a bit more code.

**Comment by EtienneBruines** [comment]
I have updated my branch to the current `origin/master`, and am now producing this benchmark:

```
BenchmarkCollisionSystem10-8  	   50000	     51255 ns/op	    1757 B/op	      86 allocs/op
BenchmarkCollisionSystem15-8  	   50000	     59302 ns/op	    1745 B/op	      86 allocs/op
BenchmarkCollisionSystem20-8  	   50000	     60601 ns/op	    1744 B/op	      86 allocs/op
BenchmarkCollisionSystem25-8  	   50000	     61777 ns/op	    1744 B/op	      86 allocs/op
BenchmarkCollisionSystem50-8  	   50000	     62944 ns/op	    1744 B/op	      86 allocs/op
BenchmarkCollisionSystem100-8 	    3000	    671916 ns/op	  134573 B/op	    3056 allocs/op
BenchmarkCollisionSystem1000-8	      50	  66592306 ns/op	12318757 B/op	  260330 allocs/op
```

Note that my CPU might be different than yours, so the raw `ns/op` should not be compared (which you probably know). 

I am investigating as to why the `allocs/op` and `B/op` becomes higher at `100` and `1000` Entities.

**Comment by EtienneBruines** [comment]
Fixed as of #88 and #89.
